### DISCUSSION TITLE
Dudas sobre el enunciado TPA 2

### DISCUSSION THREAD

**Discussion by jcastagno99** [dds-jv/foro#137] [TPA - 2020] [answered]
Buenas, tenemos problemas para interpretar los primeros puntos del enunciado : 

2) Cada ítem de un presupuesto debe estar asociado obligatoriamente a un egreso (compras). No es posible cargar presupuestos si previamente no se cargó el egreso correspondiente (esto se define para evitar que el sistema incorpore presupuestos de productos o servicios que luego no se adquieren y que no se convierta en un repositorio de presupuestos, lo que lo alejaría de su objetivo)

3) Se desea verificar las siguientes condiciones en los egresos y los presupuestos: 

a) Si la compra requiere presupuestos (para lo cual se debe tener previamente 
  configurados cuántos presupuestos son requeridos para una compra) se debe 
  verificar que efectivamente se encuentran cargada la cantidad indicada

b)  Si la compra requiere presupuestos se debe verificar que la compra haya sido 
  realizada en base a alguno de esos presupuestos

 c) Si el criterio de selección del proveedor fuera el de menor valor, debe validarse 
  que, entre los presupuestos cargados en el sistema, se haya seleccionado el de 
  menor valor. 

El punto 2 nos da a entender que un presupuesto tiene egresos (nosotros pensamos que un presupuesto tenia una lista de egresos y que cada egreso tenia una lista de items, esa iba a ser la relación presupuesto/egresos)

Pero el punto 3, en el apartado a) nos hace pensar que es la compra quien tiene presupuestos y puede tener [0,n] presupuestos (el apartado b) da a entender que la compra puede no tenerlos)

No sabemos bien como poder modelar esta relación.

**Comment by roli93** [comment]
Hola, @jcastagno99 !

Más allá de la relación que decidan a nivel objetos, que es una decisión de ustedes como equipo, conversarán con su ayudante y para la cual puede haber múltiples variantes válidas, vamos  a aclarar un poco estos requerimientos para despejar dudas:

1) En el sistema se cargan inicialmente las compras, que pueden tener o no presupuestos asociados y eso será validado más adelante.

2) Luego de haber cargado la compra, se puede cargar un presupuesto asociándolo a una compra ya existente. No es posible cargar un presupuesto y no asociarlo a ninguna compra.

3) En un instante posterior, se ejecutarán las validaciones sobre las compras. Acá se validará que si la compra requiere N presupuestos, se hayan cargado al menos N presupuestos asociados a ella. También se validará que la compra se haya efectuado en base a uno de esos presupuestos (Es decir, si los presupuestos son de una heladera de $20K, $25K y $30K, la compra no podrá ser de un televisor por $15K, por ejemplo). Finalmente, si corresponde, se hará la validación del proveedor más barato.

Nótese que, si la compra requiere presupuestos, entre el instante 1 y el 2 la misma será técnicamente "inválida" y eso es correcto, porque la validación se ejecutará después (por eso el período de gracia y que la validación no se ejecute al momento de cargar la compra, que sería lo meas natural)

Habiendo clarificado el requerimiento, la forma (referencia? uso? algún mecanismo más complejo?...) o el sentido de la asociación ("flecha") entre `Compra` y `Presupuesto` será una decisión que deberán tomar ustedes. Si aún les cuesta decidirse por una alternativa, consulten a su ayudante designado dado que probablemente la respuesta sea dependiente de como modelaron la primera entrega en su solución particular.

Saludos!

Roli

**Comment by mari967** [comment]
Hola,
Estuve siguiendo esta pregunta, me surgieron un par de dudas sobre la respuesta y el enunciado

Me genera dudas la parte del enunciado que dice *"Cada ítem de un presupuesto debe estar asociado obligatoriamente aun ​egreso​ (compras)"*. ¿Sería correcto decir que es todo el presupuesto el que está asociado a una compra y no cada uno de sus items?

¿Que una compra requiera presupuesto es algo que se decide al crear la compra y no cambia dinámicamente?

¿Los criterios de selección del presupuesto pueden ser otros, además de elegir el que tenga el menor valor?

¿Si aún no se seleccionó un presupuesto, entre los instantes 1 y 2, significa que no debería tener un proveedor seleccionado?

Gracias!

**Comment by roli93** [comment]
1) Sí, pueden asumir que la relación es a nivel **presupuesto <-> compra** y no 'item de presupuesto <-> compra'. 

2) No estamos seguros de que ese valor (el requerir presupuestos) no vaya a cambiar. Sin embargo, si cambiara luego de la validación, no es necesario revalidar automáticamente por el momento. De ser necesario, alguien volverá a correr el proceso

3) Por ahora es ese solo. Nos gustaría poder contemplar otros, pero por el momento no estamos seguros del todo de que al final vayamos a necesitarlos.

4) Si aún no se seleccionó un presupuesto, entre los instantes 1 y 2, significa solamente eso, que aún no se seleccionó. La validación no aplicará todavía. Pero si luego se seleccionara y alguien volviera a correr la validación, esto debería ser verificado.

Espero haber sido de ayuda!

Saludos!

Roli

**Comment by mari967** [comment]
Si, me quedó más claro. Muchísimas gracias 👍

**Comment by atta25** [comment]
Hola! agrego algunas consultas más:

> En el sistema se cargan inicialmente las compras, que pueden tener o no presupuestos asociados y eso será validado más adelante.

Cuando hablamos de "compras", estamos hablando de "compras concretadas" o de "posibles compras" (?)

> En un instante posterior, se ejecutarán las validaciones sobre las compras. Acá se validará que si la compra requiere N presupuestos, se hayan cargado al menos N presupuestos asociados a ella. También se validará que la compra se haya efectuado en base a uno de esos presupuestos (Es decir, si los presupuestos son de una heladera de $20K, $25K y $30K, la compra no podrá ser de un televisor por $15K, por ejemplo). Finalmente, si corresponde, se hará la validación del proveedor más barato.

Qué es lo que voy a obtener cuando se ejecuten las validaciones?
Por lo que entiendo voy a tener una colección de compras pendientes de validación y cuando ejecute las validaciones obtendré una colección de compras, que son las que cumplieron las validaciones (si tuviera sólo las 2 primeras validaciones sin la del proveedor más barato). Y si tuviese las 3 validaciones entonces obtendría una única compra que pase todas las validaciones, salvo que tenga 2 presupuestos con el mismo valor, en ese caso particular consideraría que haya al menos un presupuesto con valor mínimo y que coincida con el monto mínimo del Egreso.
Después de leer el punto 4 del enunciado quedo aún más confundido, porque interpreto que son las presupuestos los que se tienen que validar. [presupuestosPendientes] -> validar() -> [presupuestosValidados]

Cito el punto 4 del enunciado:

> Sin embargo, no se desea evitar la creación de presupuestos que incumplan las reglas anteriores,
> sino dar un tiempo prudencial para cargarlos y ajustarlos a estas reglas. Luego, en un momento
> posterior, se ejecutarán estas validaciones. Por el momento se solicita que este proceso se pueda
> ejecutar de forma manual , por consola, para todos los presupuestos pendientes de
> validación .

**Comment by BrianLuna** [comment]
Buenas! Agrego más dudas que tenemos con el equipo:

1) Documentos comerciales en presupuestos

> Se debe llevar registro de los presupuestos, tanto el total, como sus detalles y los documentos comerciales (manteniendo los datos requeridos para las operaciones de egresos).

No entendemos la función que van a tener los presupuestos, en particular lo referente a los documentos comerciales. Lo que nosotros entendemos es que el documento no es algo que esté relacionado con el presupuesto directamente, sino algo asociado al egreso que se concreta con cierto presupuesto. Por qué pide que los documentos comerciales tengan alguna relación con el presupuesto?

2) Validación de "presupuestos"

> Sin embargo, no se desea evitar la creación de presupuestos que incumplan las reglas anteriores, sino dar un tiempo prudencial para cargarlos y ajustarlos a estas reglas. Luego, en un momento posterior, se ejecutarán estas validaciones.

En varios puntos del enunciado se da a entender que lo que se valida es el presupuesto. Pero las tres validaciones que se piden en el punto 3 son sobre los egresos:

> a) Si la compra requiere presupuestos (para lo cual se debe tener previamente configurados cuántos presupuestos son requeridos para una compra) se debe verificar que efectivamente se encuentran cargada la cantidad indicada
b ) Si la compra requiere presupuestos se debe verificar que la compra haya sido realizada en base a alguno de esos presupuestos
c) Si el criterio de selección del proveedor fuera el de menor valor, debe validarse que, entre los presupuestos cargados en el sistema, se haya seleccionado el de menor valor. 

Lo único que encontramos para validar en el presupuesto en particular es que tenga un egreso asociado.

Dicho lo anterior, no entendemos si al decir: 

> [...] no se desea evitar la creación de presupuestos que incumplan las reglas anteriores, sino dar un tiempo prudencial para cargarlos y ajustarlos a estas reglas.

se están refiriendo a:

> [...] no se desea evitar la creación de **egresos** que incumplan las reglas anteriores, sino dar un tiempo prudencial para cargarlos y ajustarlos a estas reglas.

o si simplemente _**las reglas**_ son únicamente que el presupuesto tenga un egreso asociado.

3) Mensajes

> Los resultados de la validación deberán ser vistos por el sistema por parte de algunos usuarios (en una bandeja de mensajes).

Los resultados de la validación son simplemente un "egreso validado correctamente" y "egreso validado con errores"? O hay que ser más específicos respecto al motivo de validación?

Por otro lado y en referencia al mismo punto, entendemos que junto con el mensaje, es necesario que informemos el egreso validado. Es decir, si el mensaje dice solo algo como _"Egreso con cantidad de presupuestos insuficientes"_, el usuario no va a entender qué egreso en particular es el que se validó con error. Entonces nos surge la duda de si es momento ahora para preocuparnos por eso. Y en caso que lo sea, si podemos resolverlo por ejemplo teniendo un atributo _identificador_.

Muchas gracias por su tiempo!

**Comment by flbulgarelli** [comment]
> En varios puntos del enunciado se da a entender que lo que se valida es el presupuesto. Pero las tres validaciones que se piden en el punto 3 son sobre los egresos

Hola @BrianLuna, efectivamente lo que se valida es el egreso, como vos decís. Ahí cambié la redacción en de ese párrafo para evitar dudas.

**Comment by roli93** [comment]
@jcastagno99 @BrianLuna @atta25 @mari967 Hay alguna duda que sigan teniendo abierta sobre esto?

**Comment by BrianLuna** [comment]
Lo que yo consulté ya fue respondido por el ayudante.

**Comment by jcastagno99** [comment]
De mi parte no, ya quedó todo claro

**Comment by atta25** [comment]
Ídem.. Gracias!

**Comment by mari967** [comment, marked as answer]
Quedó todo claro. Muchas gracias
